src/executor.rs

```diff
@@ -25,161 +30,90 @@ pub struct Task {
-                        let out_thread = if with_stdout {
-                            let stdout = match child.stdout.take() {
-                                Some(stdout) => stdout,
-                                None => {
-                                    if let Err(e) = line_sender.send(CommandExecInfo::Error(
-                                        "taking stdout failed".to_string()
-                                    )) {
-                                        debug!("error when sending stdout error: {}", e);
-                                        break;
-                                    }
-                                    continue;
-                                }
+
+                            let task = match task {
+                                // no more tasks, exit now
+                                None => break,
+                                Some(task) => task,
                             };
-                            let line_sender = line_sender.clone();
-                            Some(thread::spawn(move|| {
-                                let mut buf = String::new();
-                                let mut buf_reader = BufReader::new(stdout);
-                                loop {
-                                    let r = match buf_reader.read_line(&mut buf) {
-                                        Err(e) => CommandExecInfo::Error(e.to_string()),
-                                        Ok(0) => {
-                                            // finished
-                                            break;
-                                        }
-                                        Ok(_) => {
-                                            // debug!("STDOUT : {:?}", &buf);
-                                            CommandExecInfo::Line(CommandOutputLine {
-                                                origin: CommandStream::StdOut,
-                                                content: TLine::from_tty(buf.trim_end()),
-                                            })
-                                        }
-                                    };
-                                    if let Err(e) = line_sender.send(r) {
-                                        debug!("error when sending stdout line: {}", e);
-                                        break;
+
+                            let child = match start_task(task, &mut command) {
+                                Err(e) => {
+                                    let response = CommandExecInfo::Error(
+                                        format!("failed to start task: {}", e)
+                                    );
+                                    match line_sender.send(response) {
+                                        Err(_) => break,
+                                        _ => continue,
                                     }
```

Comment:
block call in async

Reply:
[See above](https://github.com/Canop/bacon/pull/79#discussion_r851250263)